api/unstable/serialize.h

```diff
@@ -0,0 +1,95 @@
+ * into a transferable object to be sent off-box or stored. This transferable object can then
+ * be "deserialized" to create an s2n-tls connection object that can talk to the original peer with
+ * the same encryption keys.
+ *
+ * Note that this feature is dangerous because it provides cryptographic material from a TLS session
+ * in plaintext. Users MUST both encrypt and MAC the contents of the outputted material to provide
+ * secrecy and integrity if this material is transported off-box. DO NOT send this material off-box
+ * without a plan to do so.
+ */
+
+/* Indicates which serialized connection version will be provided. */
+typedef enum {
+    S2N_SERIALIZED_CONN_NONE = 0,
+    S2N_SERIALIZED_CONN_V1 = 1
+} s2n_serialization_version;
+
+/**
+ * Request a specific serialized connection version to be provided post-handshake.
+ *
+ * @param config A pointer to the config object.
+ * @param version The requested version.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on error.
+ */
+S2N_API int s2n_config_set_serialized_connection_version(struct s2n_config *config, s2n_serialization_version version);
+
+/**
+ * Retrieves the length of the serialized connection retrieved from s2n_connection_serialize. Should be
+ * used to allocate enough memory for the serialized connection buffer.
+ * 
+ * @note The size of the serialized connection changes based on parameters negotiated in the TLS
+ * handshake. Do not expect the size to always remain the same.
+ * 
+ * @param conn A pointer to the connection object.
+ * @param length Output parameter where the length will be written.
+ * @returns S2N_SUCCESS on success, S2N_FAILURE on error.
+ */
+S2N_API int s2n_connection_serialization_length(struct s2n_connection *conn, uint32_t *length);
+
+/**
+ *  Serializes the s2n_connection into the provided buffer.
+ * 
+ * @note You MUST call `s2n_config_set_serialized_connection_version` on the s2n_config object used
+ * to configure this connection BEFORE calling this API.
+ * @note Call `s2n_connection_serialization_length` to retrieve the amount of memory needed for the
+ * buffer parameter.
+ * @note This API will error if the handshake is not yet complete.
+ * 
+ * @param conn A pointer to the connection object.
+ * @param buffer A pointer to the buffer where the serialized connection will be written.
+ * @param buffer_length Maximum amount of data that can be written to the buffer param.
+ */
+S2N_API int s2n_connection_serialize(struct s2n_connection *conn, uint8_t *buffer, uint32_t buffer_length);
+
+/**
+ * Deserializes the provided buffer into the `s2n_connection` parameter.
+ * 
+ * @warning s2n-tls DOES NOT check the integrity of the provided buffer. s2n-tls may successfully 
+ * deserialize a corrupted buffer which WILL cause a connection failure when attempting to resume
+ * sending/receiving encrypted data. To avoid this, it is recommended to MAC the serialized 
+ * connection before sending it off-box and deserializing it.
```

Comment:
Should we also mention encryption recommendations?

Reply:
What recommendations are you thinking of?

Reply:
Something along the lines of it containing plaintext secrets, so it should be encrypted before it's sent anywhere

Reply:
I do include this note in the intro to the serialize function: https://github.com/aws/s2n-tls/pull/4493/files#diff-226b02dcaa740d80d1b4236710b61c19be9f6fd3e374da283e5290b4aadf0039R3721